---
title: "Data"
author: "XSun"
date: "2024-06-03"
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
---

# Genotype data

The data is storaged at `/project2/xinhe/xsun/cart/data`

STATs for the raw data, obtained by `bcftools stats CART_MERGED.vcf.gz > stats.txt`

| stats        |            |
|--------------|------------|
| #of samples  | 105        |
| #of variants | 85,265,672 |
| #of SNPs     | 81,712,001 |
| #of MNPs     | 249        |
| #of indels   | 3,489,092  |
| #of others   | 64,330     |

## What the origin VCF has done

File Generation and Format: The file adheres to the VCF (Variant Call Format) version 4.2, which is a standard format for storing gene sequence variations.

Source and Generation: The VCF file was generated by "Gencove SaaS v2.0.0" on the date "20/09/2023." Gencove is a platform likely used here for processing genomic data.

Reference and Imputation Panel: The sequences have been aligned to the human genome build b37_g1k v1_0_0, and an imputation panel human/b37_g1k_full_plusYMT v1_0_0 has been used. This suggests that imputation methods were applied to estimate missing genotypes.

File Content and Metadata: The header includes definitions for various metadata like sample IDs (gencovesampleid), reference allele frequencies (RAF), alternate allele frequencies (AF), and genotype qualities (GP, DS, GT). The presence of INFO and FORMAT tags suggests structured annotations related to genotypes and variant information for the genomic samples.

Merging Process: The use of bcftools_mergeCommand in the header indicates that multiple VCF files were merged into this one. The command provides a list of all the files that were merged and the specific parameters (merge -O v) used with bcftools, a tool for manipulating VCF files.

Quality Filters: Quality filters such as PASS and LOWCONF are applied, indicating that variants are filtered based on certain quality metrics. The LOWCONF filter, for example, is applied to variants where no genotype has a posterior probability (GP) greater than 0.9 (MAX(FORMAT/GP)>0.9). 

## Further QC 

1. Variant quality control:

- Removing the variants annotated as 'LOWCONF' 

- Removing the variants with MAF < 0.01

`bcftools view -i 'INFO/AF >= 0.01' -f PASS -O z -o lowconf_maf_filtered.vcf.gz CART_MERGED.vcf.gz`


| stats        |            |
|--------------|------------|
| #of samples  | 105        |
| #of variants | 1,806,562  |
| #of SNPs     | 1,662,663  |
| #of MNPs     | 2          |
| #of indels   | 143,356    |
| #of others   | 541        |

Depth coverage: 

```{r}
data <- data.table::fread("/project2/xinhe/xsun/cart/data/mediate_files/summary_depths.txt", header = FALSE, col.names = c("Chrom", "Pos", "RSID", "Ref", "Alt", "MinDepth", "MaxDepth", "AvgDepth", "Count"))
summary(data$AvgDepth)  # Provides a statistical summary of average depths
hist(data$AvgDepth, breaks = 50, main = "Distribution of Average Depth", xlab = "Average Depth")  # Histogram of average depth
```

MAF distribution after QC

```{r}
data <- data.table::fread("/project2/xinhe/xsun/cart/data/variant_ids_and_maf.txt", header = FALSE, col.names = c("snp","maf"))
hist(data$maf, breaks = 50, main = "Distribution of MAF", xlab = "MAF") 
```

2. Sample quality control:

- Computing missing genotype rates for each sample `vcftools --gzvcf lowconf_maf_filtered.vcf.gz --missing-indv --out missingness`

The max missing rate is 0.0417921%, so all samples were kept.

## Comparing with UKBB data

```{r}
load("/project2/xinhe/xsun/cart/data/process/overlap_ukbb001.rdata")
sprintf("the number of variants overlapped with UKBB variants(MAF >0.01) is %s", nrow(merged_var001))
sprintf("the fraction of variants overlapped with UKBB variants(MAF >0.01) is %s", round(nrow(merged_var001)/nrow(data),digits = 4))
```



