---
title: "Data"
author: "XSun"
date: "2024-06-03"
output:
  workflowr::wflow_html:
    code_folding: hide
    toc: true
---


```{r message=F, warnings=F}
library(VennDiagram)

library(ggplot2)

plot_histogram_with_labels <- function(data_vector, x_label = "", y_label = "Count", main="") {

  hist(data_vector, main = main,xlab = x_label, ylab = y_label)

  # Get the bin breaks and bin counts
  bins <- hist(data_vector, plot = FALSE)$breaks
  counts <- hist(data_vector, plot = FALSE)$counts

  # Add the counts to the histogram
  for (i in 1:length(counts)) {
  text(x = (bins[i] + bins[i+1]) / 2, 
       y = counts[i], 
       labels = counts[i], 
       adj = c(0.5, -0.5))
}
  
}

```

# Genotype data

The data is storaged at `/project2/xinhe/xsun/cart/data`

STATs for the raw data, obtained by `bcftools stats CART_MERGED.vcf.gz > stats.txt`

| stats        |            |
|--------------|------------|
| #of samples  | 105        |
| #of variants | 85,265,672 |
| #of SNPs     | 81,712,001 |
| #of MNPs     | 249        |
| #of indels   | 3,489,092  |
| #of others   | 64,330     |


We re-computed the allele frequency using `bcftools +fill-tags CART_MERGED.vcf.gz -Oz -o af.recomp.vcf.gz -- -t AF`


```{r fig.height=7,fig.width=14}
load("/project2/xinhe/xsun/cart/data/process/maf.raw.rdata")
plot_histogram_with_labels(data_vector = variant_raw_maf, x_label = "maf", y_label = "Count", main= "Distribution of MAF from raw data")

load("/project2/xinhe/xsun/cart/data/process/maf.rmaf.raw.plot.sampled.rdata")
print(p)

# sprintf("number of variants with MAF = 0 is %s" ,sum(variant_raw_maf==0))
# sprintf("the fraction is %s" ,sum(variant_raw_maf==0)/length(variant_raw_maf))
# 
# variant_raw_maf_non0 <- variant_raw_maf[variant_raw_maf!=0]
# plot_histogram_with_labels(data_vector = variant_raw_maf_non0, x_label = "maf", y_label = "Count", main= "maf dist for the variants with non-0 values")
```


## What the origin VCF has done

File Generation and Format: The file adheres to the VCF (Variant Call Format) version 4.2

Source and Generation: The VCF file was generated by "Gencove SaaS v2.0.0" on the date "20/09/2023."

Reference and Imputation Panel: The sequences have been aligned to the human genome build b37_g1k v1_0_0, and an imputation panel human/b37_g1k_full_plusYMT v1_0_0 has been used. This suggests that imputation methods were applied to estimate missing genotypes.

File Content and Metadata: The header includes definitions for various metadata like sample IDs (gencovesampleid), reference allele frequencies (RAF), alternate allele frequencies (AF), and genotype qualities (GP, DS, GT). The presence of INFO and FORMAT tags suggests structured annotations related to genotypes and variant information for the genomic samples.

Merging Process: The use of bcftools_mergeCommand in the header indicates that multiple VCF files were merged into this one. The command provides a list of all the files that were merged and the specific parameters (merge -O v) used with bcftools, a tool for manipulating VCF files.

Quality Filters: Quality filters such as PASS and LOWCONF are applied, indicating that variants are filtered based on certain quality metrics. The LOWCONF filter, for example, is applied to variants where no genotype has a posterior probability (GP) greater than 0.9 (MAX(FORMAT/GP)>0.9). 

## Further QC 

1. Variant quality control:

- Removing the variants annotated as 'LOWCONF' 

- Removing the variants with MAF < 0.01

`bcftools view -i 'INFO/AF >= 0.01 && INFO/AF <= 0.99' -f PASS -O z -o lowconf_maf_filtered.vcf.gz CART_MERGED.vcf.gz`

- Removing the variants on chrX

`bcftools view -r 1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22 lowconf_maf_filtered.vcf.gz -Oz -o chrxfiltered.vcf.gz`

| stats        |            |
|--------------|------------|
| #of samples  | 105        |
| #of variants | 5,190,186  |
| #of SNPs     | 4,792,775  |
| #of MNPs     | 0          |
| #of indels   | 395,870    |
| #of others   | 1,541      |


Depth coverage: 

```{r fig.height=7,fig.width=14}
data <- data.table::fread("/project2/xinhe/xsun/cart/data/process/summary_depths.txt", header = FALSE, col.names = c("Chrom", "Pos", "Ref", "Alt", "MinDepth", "MaxDepth", "AvgDepth", "Count"))
summary(data$AvgDepth)  # Provides a statistical summary of average depths
plot_histogram_with_labels(data_vector = as.numeric(data$AvgDepth), x_label = "Average Depth", y_label = "Count", main= "Distribution of Average Depth")
```

MAF distribution after QC

```{r fig.height=7,fig.width=14}
variant_cart <- data.table::fread("/project2/xinhe/xsun/cart/data/process/variants_af_raf_filtered.txt", header = FALSE, col.names = c("snp","chr","AF","RAF"))
variant_cart$MAF <- ifelse(variant_cart$AF < 0.5, variant_cart$AF, 1 - variant_cart$AF)
plot_histogram_with_labels(data_vector = variant_cart$MAF, x_label = "MAF", y_label = "Count")

nors_index_cart <- grep(pattern = "rs",x = variant_cart$snp,invert = T)
sprintf("there are %s variants without rs id", length(nors_index_cart))
print("To compare with UKBB, we removed these variants")

variant_cart <- variant_cart[-nors_index_cart,]

count_maf_over_001 <- sum(variant_cart$MAF > 0.01)
count_maf_over_005 <- sum(variant_cart$MAF > 0.05)
count_maf_over_01 <- sum(variant_cart$MAF > 0.1)
count_maf_over_02 <- sum(variant_cart$MAF > 0.2)

# Get the total number of rows
total_rows <- nrow(variant_cart)

# Create a table summarizing the counts
maf_counts <- data.frame(
  Threshold = c("MAF>0.01", "MAF>0.05", "MAF>0.1", "MAF>0.2", "Total variants"),
  Variant_Count = c(count_maf_over_001, count_maf_over_005, count_maf_over_01, count_maf_over_02, total_rows)
)

# Display the table
maf_counts

```

2. Sample quality control:

- Computing missing genotype rates for each sample `vcftools --gzvcf chrxfiltered.vcf.gz --missing-indv --out missingness`

The max missing rate is 0%, so all samples were kept.

## Comparing with UKBB data

### MAF of UKBB data

```{r fig.height=7,fig.width=14}
load("/project2/xinhe/xsun/cart/data/process/ukbb.var.rdata")
ukbb$maf <- ifelse(ukbb$allele_freq > 0.5, 1-ukbb$allele_freq, ukbb$allele_freq)
plot_histogram_with_labels(data_vector = ukbb$maf, x_label = "MAF", y_label = "Count", main = "UKBB MAF")

nors_index_ukbb <- grep(pattern = "rs",x = ukbb$id,invert = T)
sprintf("there are %s variants without rs id", length(nors_index_ukbb))
print("To compare with CarT cohort, we removed these variants")

ukbb <- ukbb[-nors_index_ukbb,]

chrx <- which(!ukbb$chrom %in% c(1:22))
sprintf("there are %s variants on chrX", length(chrx))

# Count the number of rows for each condition
count_maf_over_001 <- sum(ukbb$maf > 0.01)
count_maf_over_005 <- sum(ukbb$maf > 0.05)
count_maf_over_01 <- sum(ukbb$maf > 0.1)
count_maf_over_02 <- sum(ukbb$maf > 0.2)

# Get the total number of rows
total_rows <- nrow(ukbb)

# Create a table summarizing the counts
maf_counts <- data.frame(
  Threshold = c("MAF>0.01", "MAF>0.05", "MAF>0.1", "MAF>0.2", "Total variants"),
  Variant_Count = c(count_maf_over_001, count_maf_over_005, count_maf_over_01, count_maf_over_02, total_rows)
)

# Display the table
maf_counts
```

### Without any filtering

```{r fig.height=6,fig.width=6}

vd <- VennDiagram::venn.diagram(list(CarT = variant_cart$snp, UKBB = ukbb$id), filename = NULL)
grid::grid.draw(vd,recording = F)

```


### MAF > 0.01

```{r fig.height=6,fig.width=6}

cutoff <- 0.01

variant_cart_select <- variant_cart[variant_cart$MAF >cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]

#create_venn_diagram(set_A = nrow(ukbb_select), set_B = nrow(variant_cart_select), overlap_AB = sum(ukbb_select$id %in% variant_cart_select$snp), category_A = "UKBB", category_B = "CarT data")

vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)

```

### MAF > 0.05

```{r fig.height=6,fig.width=6}

cutoff <- 0.05

variant_cart_select <- variant_cart[variant_cart$MAF >cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]

#create_venn_diagram(set_A = nrow(ukbb_select), set_B = nrow(variant_cart_select), overlap_AB = sum(ukbb_select$id %in% variant_cart_select$snp), category_A = "UKBB", category_B = "CarT data")
vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)

```

### MAF > 0.1

```{r fig.height=6,fig.width=6}

cutoff <- 0.1

variant_cart_select <- variant_cart[variant_cart$MAF >cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]

#create_venn_diagram(set_A = nrow(ukbb_select), set_B = nrow(variant_cart_select), overlap_AB = sum(ukbb_select$id %in% variant_cart_select$snp), category_A = "UKBB", category_B = "CarT data")
vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)
```

### MAF > 0.2

```{r fig.height=6,fig.width=6}

cutoff <- 0.2

variant_cart_select <- variant_cart[variant_cart$MAF > cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]


vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)
```


## More investigation

The number of common variants in our data is still somewhat lower than expected. Ex. for high frequency variants (MAF > 0.2), about half of UKB variants are missing in our data. So weâ€™ll do some more investigation to see why.

We first checked if some of the UKB common variants overlap with the CarT variants with MAF < 0.2 

```{r fig.height=6,fig.width=6}

cutoff <- 0.2

variant_cart_select <- variant_cart[variant_cart$MAF < cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]

vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)
```

There are over 74k variants have low MAF in cart cohort. We checked the distribution of these MAFs

```{r fig.height=6,fig.width=6}
overlap <- variant_cart_select[variant_cart_select$snp %in% ukbb_select$id,]
hist(overlap$MAF, main ="CarT MAF of the common UKB variants (overlaps)")

overlap$RMAF <- ifelse(overlap$RAF > 0.5, 1-overlap$RAF, overlap$RAF)
hist(overlap$RMAF, main ="reference MAF of the common UKB variants (overlaps)")
```

If we filter the CarT variants by reference allele frequency, we have 

```{r fig.height=6,fig.width=6}

cutoff <- 0.2

variant_cart$RMAF <- ifelse(variant_cart$RAF > 0.5, 1-variant_cart$RAF, variant_cart$RAF)
variant_cart_select <- variant_cart[variant_cart$RMAF > cutoff,]
ukbb_select <- ukbb[ukbb$maf > cutoff,]


vd <- VennDiagram::venn.diagram(list(CarT = variant_cart_select$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)
```

The number of overlaps doesn't increase. 

Then, we checked if some of the UKB common variants overlap with the CarT variants that were filtered out by 'LOWCONF'

```{r fig.height=6,fig.width=6}

load("/project2/xinhe/xsun/cart/data/process/cart_var_lowconf.rdata")

ukbb_select <- ukbb[ukbb$maf > cutoff,]

vd <- VennDiagram::venn.diagram(list(CarT_lowconf = cart_raw_nochrx_lowconf$snp, UKBB = ukbb_select$id), filename = NULL)
grid::grid.draw(vd,recording = F)
```


